File: time_travel_to_programming_languages/src/bern.erl
% Skriven av Hugo Larsson Wilhelmsson och Erik Smit

% Skapar en modul bern, som räknar ut Bernoullital
-module(bern).
% Funktionerna som är synliga utanför modulen
-export([bernoulli/1, bern/0]).

% Funktion för att beräkna ett Bernoullital N
% N kan inte vara mindre än 0
bernoulli(N) when N >= 0 ->
    % Returnerar det sista talet i listan
    lists:last(
        % Skapar en lista med N-1 element där M är 1 till N 
        % och Acc är i början listan med ett element, 1.
        % fun(M, Acc) är en funktion som genomförs på varje element
        lists:foldl(
            fun(M, Acc) ->
                % Anropar hjälpfunktionen calc_b för att fortsätta algoritmen
                Bm = calc_b(M, Acc),
                % Lägger till Bm i listan Acc
                Acc ++ [Bm]
            end, [1], lists:seq(1, N)
        )
    ).

% Funktion för att beräkna binomialkoefficienten
binom(N, K) ->
    R = 1,
    % Loopar igenom koden fun(I, Acc) K-1 gånger där N har värdet 1 till K. 
    % Acc börjar med värdet R = 1 och uppdateras med det nya värdet i varje 
    % loop och returneras i slutet
    lists:foldl(
        fun(I, Acc) ->
            Acc * (N - I + 1) / I
        end, R, lists:seq(1, K)
    ).

% Hjälpfunktion för att beräkna B[m] enligt algoritmen
calc_b(M, B) ->
    % Loopar M-1 gånger där M är 0 till M-1. Acc är i början 0. 
    % Värdet på Acc ändras i varje loop.
    Bm = lists:foldl(
        fun(K, Acc) ->
            % Anropar Binom och multiplicerar det med B[k].
            % K+1 är index i B och vi tar K+1 eftersom listor 
            % börjar på index 1 i Erlang
            Acc - binom(M + 1, K) * lists:nth(K+1, B)
        end, 0, lists:seq(0, M-1)),
    % Sista steget i algoritmen. Funktionen returnerar sedan Bm
    Bm / (M + 1).

% Funktion för att skriva ut de 20 första Bernoullitalen
bern() ->
    % Ändra N för att skriva ut de N första talen
    N = 20,
    [io:format("B[~p]: ~s~n", [X, format_bernoulli(bernoulli(X))]) || X <- lists:seq(0, N)].

format_bernoulli(Number) ->
    io_lib:format("~.9f", [float(Number)]).

File: time_travel_to_programming_languages/src/bern.rs
// Skriven av Hugo Larsson Wilhelmsson och Erik Smit

// Funktion för att beräkna Bernoullitalen
fn bernoulli() {
    // Skapar en vektor, bernoulli_numbers, som är mutable. Vi säger att den har ett element, 1f64 = 1 i 64 bitar
    // ! står innan makron
    let mut bernoulli_numbers = vec![1f64];

    // Algoritmen från uppgiften
    // Loopar från 1 till 21
    // Ändra 21 till annat önskas N+1 om så önskas, ändra även nere i printen
    for m in 1..=21 {
        // Lägger till elementet 0f64 = 0 i vektorn bernoulli_numbers
        bernoulli_numbers.push(0f64);
        // Loop från 0 till m-1
        for k in 0..m {
            bernoulli_numbers[m] -= binomial(m + 1, k) * bernoulli_numbers[k];
        }
        // Konverterar m till f64 och adderar med 1f64 = 1
        bernoulli_numbers[m] /= m as f64 + 1f64;
    }

    // Printar de 21 första Bernoullitalen (0-20) med 9 decimaler
    // Ändra 20 till annat önskat N om så önskas, ändra även i loopen ovan
    for n in 0..=20 {
        println!("B[{}] = {:.9}", n, bernoulli_numbers[n]);
    }
}

// Funkttion för att beräkna binomialkoefficienten
fn binomial(n: usize, k: usize) -> f64 {
    // Initierar variabeln r som är mutable
    let mut r = 1f64;
    // Loopar från 1 till k
    for i in 1..=k {
        r *= (n - i + 1) as f64 / i as f64;
    }
    r
}

// main-funktionen som anropar bernoulli()
fn main() {
    bernoulli();
}